src/app/ReadHandler.cpp

```diff
@@ -742,5 +748,34 @@ CHIP_ERROR ReadHandler::RefreshSubscribeSyncTimer()
 
     return CHIP_NO_ERROR;
 }
+
+void ReadHandler::ResetPathIterator()
+{
+    mAttributePathExpandIterator = AttributePathExpandIterator(mpAttributeClusterInfoList);
+    mAttributeEncoderState       = AttributeValueEncoder::AttributeEncodeState();
+}
+
+void ReadHandler::SetDirty(const ClusterInfo * apAttributeChanged)
+{
+    ConcreteAttributePath path;
+    mDirtyTick     = InteractionModelEngine::GetInstance()->GetReportingEngine().GetDirtyTick();
+    mOverrideDirty = true;
+
+    // We won't reset the path iterator for every SetDirty call to reduce the number of full report data.
+    // The iterator will be reset after finished each report session.
+    if (apAttributeChanged == nullptr ||
+        (mAttributePathExpandIterator.Get(path) && path.mEndpointId == apAttributeChanged->mEndpointId &&
+         path.mClusterId == apAttributeChanged->mClusterId) ||
```

Comment:
There is a bug here - while this logic works great for Reads, in the case of steady-state report generation that utilizes the global dirty set to further mask the request expanded path set, you need to actually reset the iterator all the way back to the beginning. Without that, you'll actually miss gather attributes who are considered newer than the last report time of the read handler, but we'll never actually iterate over them since our iterator has already gone past those.

Reply:
With the help of the timestamp, I think they should be included in the next report message, that why ReadHandler.cpp:260 should be the timestamp of the start of the report series.

In this case, any attribute change after the start of the current report will be included in the next report.

Consider this case: we have finished the priming report, then we added a new endpoint -- this will result in a large number of chunks, and another attribute keeps changing, then we should not use a aggressive reset strategy here too.

Reply:
> In this case, any attribute change after the start of the current report will be included in the next report.

I don't think this is optimal. Those attributes **have** to be bundled into some report, so it's better to do that in the current report if one is already in flight. Otherwise, you may have to either wait for minimum reporting interval to generate that report, as well as potentially making that report in-efficient (you may have had space in the current reporting series to stuff that in).

Reply:
I've come around to agreeing with @erjiaqing , that over-eagerly bundling dirty attributes in the current reporting series will cause issues with high rate of change attributes.